### PR TITLE
gadget/gadget.go: LaidOutSystemVolumeFromGadget -> LaidOutVolumesFromGadget

### PR DESCRIPTION
**Pull request by anonymouse64**
The new variant of this function returns the specific boot/system volume, as
well as all volumes that have been laid out so that they can be matched with
specific physical devices during UC20+ install mode for multi-volume gadget
asset updates.

Based on top of https://github.com/snapcore/snapd/pull/10790

### CODE REVIEW

File: gadget/gadget.go [hunk 1]
```diff
@@ -962,22 +964,28 @@ func LaidOutSystemVolumeFromGadget(gadgetRoot, kernelRoot string, model Model) (
...
 		for _, structure := range vol.Structure {
-			// use the system-boot role to identify the system volume
 			if structure.Role == SystemBoot {
-				pvol, err := LayoutVolume(gadgetRoot, kernelRoot, vol, constraints)
-				if err != nil {
-					return nil, err
-				}
-
-				return pvol, nil
+				system = lvol
```
Comment (mardy) [inline gadget/gadget.go:978]: Should we check that `system == nil` before doing this assignment? I guess we expect to find a single system volume anyway, but is it possible (maybe due to a gadget misconfiguration) that there are more than one?

If that is possible, the old code here would return the *first* one found, while with the new code we'd be returning the *last* one.

**Review by mardy — APPROVED** [review]
LGTM, with a question :-)

File: gadget/gadget.go (hunk 1 above)
Comment (anonymouse64) [reply gadget/gadget.go:978]: This actually is impossible, as the gadget.yaml must have passed validation above with ReadInfoAndValidate, but I some paranoia is justified as we have had bad gadget bugs before where folks couldn't refresh/update their gadget, so explicitly failing with an internal error here feels like the right thing to do

File: gadget/gadget.go [hunk 2]
```diff
@@ -938,22 +938,24 @@ func IsCompatible(current, new *Info) error {
 	return nil
 }
 
-// LaidOutSystemVolumeFromGadget takes a gadget rootdir and lays out the
-// partitions as specified. It returns one specific volume, which is the volume
-// on which system-* roles/partitions exist, all other volumes are assumed to
-// already be flashed and managed separately at image build/flash time, while
-// the system-* roles can be manipulated on the returned volume during install
+// LaidOutVolumesFromGadget takes a gadget rootdir and lays out the partitions
+// as specified. It returns the specific volume on which system-*
```
Comment (bboozzoo) [inline gadget/gadget.go:942]: // on all volumes as specified.

File: gadget/gadget_test.go [hunk 3]
```diff
@@ -111,8 +111,6 @@ volumes:
         type: bare
         size: 623000
         offset: 0
-        content:
```
Comment (bboozzoo) [inline gadget/gadget_test.go:114]: I think I'd leave the content entry and just create a file with some data in the test such that we encode the fact that even wit multi volume gadget the content of volumes other than system must exist in the gadget directory.

**Review by bboozzoo — COMMENTED** [review]
LGTM

**Review by mardy — APPROVED** [review]

File: gadget/gadget_test.go (hunk 3 above)
Comment (anonymouse64) [reply gadget/gadget_test.go:114]: Good point I will add this back in

Comment (anonymouse64) [reply gadget/gadget_test.go]: done

File: gadget/gadget.go (hunk 2 above)
Comment (anonymouse64) [reply gadget/gadget.go]: added

**Review by bboozzoo — APPROVED** [review]
Thanks

**Pull request merged**